src/mygrad/_graph_tracking.py

```diff
@@ -0,0 +1,52 @@
+"""
+Provides user interface for suspending computational graph tracking and back-propagation
+"""
+from functools import wraps
+from typing import Callable
+
+import numpy as np
+
+__all__ = ["no_autodiff"]
+
+
+# If `False`, suspends all computational graph tracking and backprop
+TRACK_GRAPH = True  # type: bool
+
+
+class NoAutoDiff:
+    """ Serves as a context manager and decorator for suspending
+    all computational graph tracking."""
+
+    def __enter__(self):
+        """Suspends graph-tracking"""
+        global TRACK_GRAPH
+        TRACK_GRAPH = False
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Restores graph-tracking"""
+        global TRACK_GRAPH
+        TRACK_GRAPH = True
```

Comment:
You need to track the stack depth, or the following patterns are broken:

```python
with no_autodiff:
	# autodiff disabled here

    with no_autodiff:
		pass # disabled here too

	# whoops, it's re-enabled here!
```

and once you're dealing in decorated functions it's pretty easy to accidentally nest calls like this.

```suggestion
    def __enter__(self):
        """Suspends graph-tracking"""
        global TRACK_GRAPH
        TRACK_GRAPH = False
        depth += 1  # global variable, self.depth attribute, whatever

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Restores graph-tracking"""
        global TRACK_GRAPH
        assert depth >= 1
        depth -= 1
        TRACK_GRAPH = depth == 0
```

Reply:
Great catch... I can only hope that I would have thought about this before this saw the light of day 😬